### ISSUE TITLE
Using generators instead of iterators for sampling (DataLoader)

### ISSUE THREAD

**Issue by bobbens** [pytorch/pytorch#683]
Instead of using iterators, I want to use generators (infinite iterators) for sampling with DataLoader. For this purpose I have wrapper a generator in an iterator class, and made ```__len__``` return a very large number. While this does work, it has two issue:

 1. When you hit ```__len__``` it'll stop iterating and you'll not necessarily ned with a full batch.
 2. DataLoader uses multiple processes, not threads, for loading, so the generator gets copied, and since ```__getitem__``` ignores the index (it is randomly generating with the generator), all the processes have the same random seed so you get the same output for all the generators in the batch.

So while I understand this is a fairly low priority issue, since I assume most people are fine with using iterators, the easiest way would be to add a new class DataGenerator, like DataLoader, except it does not have ```__len__``` and initializes and runs separate iterators in each process. Although it should also be possible to say add another parameter to DataLoader and handle it there. I can do a pull request if interested.

**Comment by apaszke** [comment]
I don't think we really need to add any special support for it. Is the generation process so expensive that you have to offload the work to other processes? Why do you want your iterators to never end? Don't you want to mark the end of an epoch anyhow? It won't hurt you to restart the loop from time to time. Also, both of your issues can be easily solved:

1. Using `2**64` as length should be pretty much the same as infinity. Assuming you can process even millions of samples per second it will be enough to keep it running for (584 942) years.
2. You can use the index to reset the generator seed, or do it once based on `os.getpid()`.

I'm not sure what kind of additional argument you'd like to add to the `DataLoader`. Can you elaborate on that?

**Comment by bobbens** [comment]
Yes, it is very expensive. Yes, I don't want the iterators to end. In fact, I am not using a single iterator, but multiple at the same time (from different sources), and have to have constant batch sizes, which is why if they reach the end it is troublesome.

 1. Yes, I am doing something like that. However, it does feel "unclean" from the code point of view.
 2. I have inserted a numpy.random.seed() in the generator itself that I trigger at the sampling start, which does work, but again is rather hackish.

From my point of view, I have it all working, although it does feel hackish and I wouldn't be surprised to see it break when the DataLoader internals slightly change. So  if you deem this not an issue (or feature bloat), feel free to close the issue.

Easiest solution would be to have a parameter to tell DataLoader it is a generator, not an iterator, and have it use for example ```__call__``` instead of ```__iter__``` in that case. That would just generate a new sample. On the random issue, either have all the processes automatically generate new random seeds, or add an option to give the worker threads a "initialization" function, which would be run when they are created. This would possibly allow other usages, while also allowing the user to randomize the seeds.

**Comment by soumith** [comment]
For now we should change iterators to be generators. The use-cases aren't common enough.
I think having infinite generators can be put in a utility library outside of the core.

**Comment by apaszke** [comment]
I don't think it's hackish, I'd say it's quite standard. I doubt you'll deplete an iterator of such length on any common hardware, for years to come. I also don't think it's going to break if the `DataLoader` internals change - it conforms to the dataset API, so it should work with all future versions too (unless we change that interface, but I doubt it).

**Comment by apaszke** [comment]
A worker initialization function sounds useful though. We could add that.

**Comment by blukee** [comment]
I run into a similar situation that it would require to sample a dataset based on probability to get evenly distributed training data from a imbalanced distribution. With the current implementation, I would need to process the training data before hand. If it's a generator, I can sample the data on the fly.
